rusoto/core/src/client_inner.rs

```diff
@@ -23,37 +23,61 @@ impl<P, D> ClientInner<P, D> {
 
 impl<P: ProvideAwsCredentials, D: DispatchSignedRequest> ClientInner<P, D> {
     /// Fetch credentials, sign the request and dispatch it.
-    pub fn sign_and_dispatch<E>(&self, request: SignedRequest) -> SignAndDispatch<P, D, E> {
+    pub fn sign_and_dispatch<F: IntoFuture>(&self, request: SignedRequest, response_handler: fn(HttpResponse) -> F) -> SignAndDispatch<P, D, F> {
         SignAndDispatch {
             credentials_provider: self.credentials_provider.clone(),
             dispatcher: self.dispatcher.clone(),
+            response_handler: response_handler,
             state: SignAndDispatchState::Lazy { request: request },
-            _phan: PhantomData
+            timeout: None
         }
     }
 }
 
-pub struct SignAndDispatch<P: ProvideAwsCredentials, D: DispatchSignedRequest, E> {
+pub trait ClientInnerFuture: Future {
```

Comment:
This trait allows us later to box a `SignAndDispatch` and get rid of the P, D, F type parameters by using a trait object instead.